beacon_node/http_api/src/beacon_proposer_cache.rs

```diff
@@ -0,0 +1,168 @@
+
+        // If the head epoch is more than `EPOCHS_TO_SKIP` in the future, just build the cache at
+        // the epoch of the head. This prevents doing a massive amount of skip slots when starting
+        // a new database from genesis.
+        let epoch = {
+            let epoch_now = chain
+                .epoch()
+                .unwrap_or_else(|_| chain.spec.genesis_slot.epoch(T::EthSpec::slots_per_epoch()));
+            let head_epoch = head_block.slot.epoch(T::EthSpec::slots_per_epoch());
+            if epoch_now > head_epoch + EPOCHS_TO_SKIP {
+                head_epoch
+            } else {
+                epoch_now
+            }
+        };
+
+        Self::for_head_block(chain, epoch, head_root, head_block)
+    }
+
+    /// Create a new cache that contains the shuffling for `current_epoch`,
+    /// assuming that `head_root` and `head_block` represents the most recent
+    /// canonical block.
+    fn for_head_block<T: BeaconChainTypes>(
+        chain: &BeaconChain<T>,
+        current_epoch: Epoch,
+        head_root: Hash256,
+        head_block: ProtoBlock,
+    ) -> Result<Self, BeaconChainError> {
+        let _timer = metrics::start_timer(&metrics::HTTP_API_BEACON_PROPOSER_CACHE_TIMES);
+
+        let mut head_state = chain
+            .get_state(&head_block.state_root, Some(head_block.slot))?
+            .ok_or_else(|| BeaconChainError::MissingBeaconState(head_block.state_root))?;
+
+        // We *must* skip forward to the current epoch to obtain valid proposer
+        // duties. We cannot skip to the previous epoch, like we do with
+        // attester duties.
+        while head_state.current_epoch() < current_epoch {
+            // Skip slots until the current epoch, providing `Hash256::zero()` as the state root
+            // since we don't require it to be valid to identify producers.
+            per_slot_processing(&mut head_state, Some(Hash256::zero()), &chain.spec)?;
+        }
+
+        let proposers = current_epoch
+            .slot_iter(T::EthSpec::slots_per_epoch())
+            .map(|slot| {
+                head_state
+                    .get_beacon_proposer_index(slot, &chain.spec)
+                    .map_err(BeaconChainError::from)
+                    .and_then(|i| {
+                        let pubkey = chain
+                            .validator_pubkey(i)?
+                            .ok_or_else(|| BeaconChainError::ValidatorPubkeyCacheIncomplete(i))?;
+
+                        Ok(ProposerData {
+                            pubkey: PublicKeyBytes::from(pubkey),
+                            slot,
+                        })
+                    })
+            })
+            .collect::<Result<_, _>>()?;
+
+        let epoch_boundary_slot = head_state
+            .current_epoch()
+            .start_slot(T::EthSpec::slots_per_epoch());
+        let epoch_boundary_root = if head_state.slot >= epoch_boundary_slot {
```

Comment:
The whole code block is:

```rust
        let epoch_boundary_slot = head_state
            .current_epoch()
            .start_slot(T::EthSpec::slots_per_epoch());
        let epoch_boundary_root = if head_state.slot >= epoch_boundary_slot {
            head_root
        } else {
            *head_state.get_block_root(epoch_boundary_slot)?
        };
```

I think this is correct as is, `epoch_boundary_slot` is always equal to or less than `state.slot` and doing `state.get_block_root()` will correctly account for skip slots.

Reply:
> epoch_boundary_slot is always equal to or less than state.slot and doing state.get_block_root() will correctly account for skip slots.

I agree with this part, but not the conclusion :sweat_smile: This observation implies `head_state.slot >= epoch_boundary_slot` will always evaluate to `true`, which means `epoch_boundary_root` will always be set to `head_root`, which is wrong.

Given that `head_state.slot < epoch_boundary_slot` is impossible, I think it should be an `==`, like:

```rust
        let epoch_boundary_slot = head_state
            .current_epoch()
            .start_slot(T::EthSpec::slots_per_epoch());
        let epoch_boundary_root = if head_state.slot == epoch_boundary_slot {
            head_root
        } else {
            *head_state.get_block_root(epoch_boundary_slot)?
        };
```

This might be a moot point anyway, if we switch to the last slot of the previous epoch, as per https://github.com/sigp/lighthouse/pull/1569/files#r494713541. Do you think that change is worth making?

Reply:
The changes in 5be8690 should make this irrelevant, thanks!